### DISCUSSION TITLE
Reading in geospatial information from grib file to convert to raster

### DISCUSSION THREAD

**Discussion by shawncrawley** [corteva/rioxarray#196] [Q&A] [answered]
This is somewhere between an idea and a question, with where they meet being a possible bug report. The question would be "Does rioxarray already support grib to some degree?" If the answer is yes, then I must have a grib whose format presents a case that is not yet supported, since I cannot get the following to work:

1. I download https://noaa-nbm-grib2-pds.s3.amazonaws.com/blend.20201216/12/core/blend.t12z.core.f001.co.grib2 as "test.grib2"
2. I run the following code:
```python
import xarray as xr
import rioxarray
ds = xr.open_dataset('test.grib2', engine='cfgrib', engine='cfgrib', backend_kwargs={'filter_by_keys': {'stepType': 'accum', 'typeOfLevel': 'surface', 'totalNumber': 'undef'}})
ds.tp.rio.to_raster('test.tif')
```
3. I get the following output:
```bash
NotGeoreferencedWarning: The given matrix is equal to Affine.identity or its flipped counterpart. GDAL may ignore this matrix and save no geotransform without raising an error. This behavior is somewhat driver-specific.
```
4. I load the tif into ArcGIS Pro and it placed as a tiny spec off the coast of Africa, rather than covering CONUS.

I even tried to explicitly set the crs with `ds.rio.set_crs('+proj=lcc +lat_0=25 +lon_0=265 +lat_1=25 +lat_2=25 +x_0=0 +y_0=0 +R=6371200 +units=m +no_defs')` as part of step 2 above before running `to_raster()`, but get the same subsequent results.

I am able to accomplish the above easily with gdal, as follows:
```bash
gdal_translate -b 46 test.grib2 test.tif
```
I would truly prefer to be able to use rioxarray rather than having to go the gdal route. Namely, because I will be doing in-memory raster math on multiple grib2 files using xarray before writing the resulting grid to raster. This would be extremely straight-forward using rioxarray, if possible.

**Comment by snowman2** [comment, marked as answer]
If you use [rioxarray.open_rasterio](https://corteva.github.io/rioxarray/stable/rioxarray.html#rioxarray-open-rasterio) it works as it uses GDAL under the hood.

```python
import rioxarray

xds = rioxarray.open_rasterio("blend.t12z.core.f001.co.grib2")
xds.attrs["long_name"] = xds.long_name[45]  # long name is 0 indexed and band is 1 indexed
xds = xds.sel(band=46)
xds.rio.to_raster('test.tif')
```

When opening up GRIB files, the spatial information isn't always read in by the software you are using. The GRIB driver in GDAL handles that for you. This logic is also included in the [pygrib](https://github.com/jswhit/pygrib) package IIRC. However, I am not sure about `cfgrib`.

> I even tried to explicitly set the crs with ds.rio.set_crs

I would recommend `rio.write_crs` as it persists better. Also, you are missing your transform and would need to figure out what that is and add that as well with `rio.write_transform`.

**Reply by shawncrawley** [reply]
Thanks, this does just what I need! I appreciate your help.
